pkg/lbtransport/transport.go

```diff
@@ -0,0 +1,140 @@
+package lbtransport
+
+import (
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/observatorium/observable-demo/pkg/conntrack"
+	"github.com/observatorium/observable-demo/pkg/runutil"
+	"github.com/pkg/errors"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Metrics struct {
+	successes *prometheus.CounterVec
+	failures  *prometheus.CounterVec
+	duration  *prometheus.HistogramVec
+
+	dialerMetrics *conntrack.DialerMetrics
+}
+
+func NewMetrics(reg prometheus.Registerer) *Metrics {
+	var m Metrics
+
+	m.successes = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: "lbtransport",
+		Name:      "rt_failures_total",
+		Help:      "Total number of cache failures against storage.",
+	}, []string{"operation"})
+	m.failures = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Subsystem: "lbtransport",
+		Name:      "rt_failures_total",
+		Help:      "Total number of cache failures against storage.",
+	}, []string{"address"})
+	m.dialerMetrics = conntrack.NewDialerMetrics(reg)
+	//m.duration = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+	//	Subsystem: "lbtransport",
+	//	Name:      "picking_duration_seconds",
+	//	Help:      "Time it took to perform cache operation.",
+	//	Buckets:   []float64{0.01, 0.1, 0.3, 0.6, 1, 3, 6, 9, 20, 30, 60, 90, 120, 240, 360, 720},
+	//}, []string{"operation"}) // TODO: Failure duration vs success duration.
+	// Not sure where duration can be useful here (:
+
+	if reg != nil {
+		reg.MustRegister(
+			m.successes,
+			m.failures,
+			m.duration,
+		)
+	}
+	return &m
+}
+
+type Transport struct {
+	discovery Discovery
+	picker    TargetPicker
+
+	metrics *Metrics
+
+	parent http.RoundTripper
+}
+
+func NewLoadBalancingTransport(discovery Discovery, picker TargetPicker, metrics *Metrics) *Transport {
+	return &Transport{
+		discovery: discovery,
+		picker:    picker,
+		metrics:   metrics,
+		parent: &http.Transport{
+			Proxy: http.ProxyFromEnvironment,
+			DialContext: conntrack.NewInstrumentedDialContextFunc(
+				(&net.Dialer{
+					Timeout:   10 * time.Second,
```

Comment:
Just to be good samaritans, we should make everything parametric and configurable.

Reply:
For demo it's fine. Maybe Todo comment would be enough.